types/wlr_output.c

```diff
@@ -188,94 +188,172 @@ void wlr_output_set_position(struct wlr_output *output, int32_t lx,
 	}
 }
 
-static bool set_cursor(struct wlr_output *output, const uint8_t *buf,
-		int32_t stride, uint32_t width, uint32_t height, int32_t hotspot_x,
-		int32_t hotspot_y) {
-	if (output->impl->set_cursor
-			&& output->impl->set_cursor(output, buf, stride, width, height,
-				hotspot_x, hotspot_y, true)) {
-		output->cursor.is_sw = false;
-		return true;
+void wlr_output_init(struct wlr_output *output, struct wlr_backend *backend,
+		const struct wlr_output_impl *impl) {
+	output->backend = backend;
+	output->impl = impl;
+	wl_list_init(&output->modes);
+	output->transform = WL_OUTPUT_TRANSFORM_NORMAL;
+	output->scale = 1;
+	wl_list_init(&output->cursors);
+	wl_signal_init(&output->events.frame);
+	wl_signal_init(&output->events.swap_buffers);
+	wl_signal_init(&output->events.resolution);
+	wl_signal_init(&output->events.destroy);
+}
+
+void wlr_output_destroy(struct wlr_output *output) {
+	if (!output) {
+		return;
 	}
 
-	wlr_log(L_INFO, "Falling back to software cursor");
+	wl_signal_emit(&output->events.destroy, output);
 
-	output->cursor.is_sw = true;
-	output->cursor.width = width;
-	output->cursor.height = height;
+	struct wlr_output_mode *mode, *tmp_mode;
+	wl_list_for_each_safe(mode, tmp_mode, &output->modes, link) {
+		free(mode);
+	}
+	wl_list_remove(&output->modes);
+	if (output->impl && output->impl->destroy) {
+		output->impl->destroy(output);
+	} else {
+		free(output);
+	}
+}
 
-	if (!output->cursor.renderer) {
-		output->cursor.renderer = wlr_gles2_renderer_create(output->backend);
-		if (!output->cursor.renderer) {
-			return false;
-		}
+void wlr_output_effective_resolution(struct wlr_output *output,
+		int *width, int *height) {
+	// TODO: Scale factor
+	if (output->transform % 2 == 1) {
+		*width = output->height;
+		*height = output->width;
+	} else {
+		*width = output->width;
+		*height = output->height;
 	}
+}
 
-	if (!output->cursor.texture) {
-		output->cursor.texture =
-			wlr_render_texture_create(output->cursor.renderer);
-		if (!output->cursor.texture) {
-			return false;
-		}
+void wlr_output_make_current(struct wlr_output *output) {
+	output->impl->make_current(output);
```

Comment:
Do we check if it has the interface methods? Seems inconsistent with destroy.

Reply:
This is the old code, unchanged. I can add the check if needed.

Reply:
There are a lot of other functions that are mandatory for outputs (`enable`, `transform`, `swap_buffers`...) so I'm going to leave it as is, unless there's a reason we should make all of these optional.

Reply:
I kind of wish there was a way to check if the required methods are there before we get to them. Wayland doesn't do this and it led to some crashes early on. I'm thinking of someone who is new to the library (which will be everyone at some point).

Reply:
That's true, a segfault should definitely not happen. Maybe make the `create` function fail and log the error?

Reply:
Add assertions. It's a library user error that needs to be fixed right away.

Reply:
(At least) the Wayland backend is missing `enable`. I wonder what's the best idea:
1. Implement `enable` for Wayland, `enable(false)` destroys the window
2. Make `enable` optional, or implement a no-op `enable` function in the Wayland backend

Reply:
Make enable optional, don't call it if it's null (no-op).